### PR TITLE
Add test docs and fix various issues

### LINKED ISSUE #555: rename db::AddTags to db::AddSimpleTags
# Feature request:

We have started referring to non-compute items in the DataBox as simple items, so we should update our naming conventions to reflect this.

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### LINKED ISSUE #591: rename swallow to expand_pack
Feature request:  find all uses

### LINKED ISSUE #601: A function that returns the first argument of a parameter pack would be useful
# Feature request:

I'd be nice to have a function that given a parameter pack, returns just the first argument. 

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Desired feature:

- [ ] Detail 1
- [ ] Detail 2
- [ ] Detail 3

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Adds documentation to the main page and for how to write tests
- Loosen tolerance in Einsum test that currently sometimes fails
- Rename swallow to expand_pack (closes #591 )
- Rename AddTags -> AddSimpleTags and AddComputeItemsTags -> AddComputeTags (closes #555 )
- Add `get_first_argument` which returns the first argument of a parameter pack (closes #601 )

The rendered documentation is here: http://gh.nilsdeppe.com/spectre/pr_617/

### Types of changes:

- [x] Bugfix
- [x] New feature

### Component:

- [x] Code
- [x] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

### CODE REVIEW

File: docs/MainSite/Main.md [hunk 1]
```diff
@@ -21,3 +21,30 @@ Introduction {#intro_sec}
...
+Navigating the Documentation {#navigate_documentation_sec}
+-----
+
+The SpECTRE documentation is organized into tutorials, developer guides, groups
+of related code, namespaces, and by files for easier navigation. For someone
+looking to run simulations with SpECTRE we recommend starting with the
+\ref tutorials "tutorials", which are located under the `Documentation` menu
+item at the top. The tutorials are designed to get users up and running with a
+simulation, and analyzing and visualizing the output. For people looking to
+contribute to SpECTRE there are a tutorials under the \ref dev_guide "Dev Guide"
```
Comment (nilsdeppe) [inline docs/MainSite/Main.md]: there are ~a~ tutorials

File: docs/DevGuide/WritingTests.md [hunk 2]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
 \endcond
 # Writing Unit Tests {#writing_unit_tests}
 
-## Testing Failure Cases
+Unit tests are placed in the appropriate subdirectory of `tests/Unit`, which
+mirrors the directory hierarchy of `src`. The tests are all compiled into
+individual libraries to keep link time of testing executable down. Typically
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: executables (plural)

File: docs/DevGuide/WritingTests.md [hunk 3]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+there should be one test library for each production code library. For example,
+we have a `DataStructures` library and a `Test_DataStructures` library. When
+adding a new test there are several scenarios that can occur, which are outlined
+below.
+
+- You are adding a new source file to an existing test library:<br>
+  If you are adding a new source file in a directory that already has a
+  `CMakeLists.txt` simply create the source file, which should be name
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: name -> named

File: docs/DevGuide/WritingTests.md [hunk 4]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+  file.<br>
+  If you are adding a new source file to a library but want to place it in a
+  subdirectory you must first create the subdirectory. To provide a concrete
+  example, say you are adding the directory `TensorEagerMath` to
+  `tests/Unit/DataStructures`. After creating the directory you must add a call
+  to `add_subdirectory(TensorEagerMath)` to
+  `tests/Unit/DataStructures/CMakeLists.txt` *before* the call to
+  `add_test_library` and *after* the `LIBRARY_SOURCES` are set. Next add the
+  file `tests/Unit/DataStructures/TensorEagerMath/CMakeLists.txt`, which should
+  add the new source files by call
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: call -> calling `set`, e.g.

File: docs/DevGuide/WritingTests.md [hunk 5]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+individual libraries to keep link time of testing executable down. Typically
+there should be one test library for each production code library. For example,
+we have a `DataStructures` library and a `Test_DataStructures` library. When
+adding a new test there are several scenarios that can occur, which are outlined
+below.
+
+- You are adding a new source file to an existing test library:<br>
+  If you are adding a new source file in a directory that already has a
+  `CMakeLists.txt` simply create the source file, which should be name
+  `Test_ProductionFileBeingTested.cpp` and add that to the `LIBRARY_SOURCES` in
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: `ProductionCodeFileBeingTest` make consistent

File: docs/DevGuide/WritingTests.md [hunk 6]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+  file `tests/Unit/DataStructures/TensorEagerMath/CMakeLists.txt`, which should
+  add the new source files by call
+  ```
+  set(LIBRARY_SOURCES
+      ${LIBRARY_SOURCES}
+      Test_ProductionCodeFileBeingTest.cpp
+      PARENT_SCOPE)
+  ```
+  The `PARENT_SCOPE` flag tells CMake to make the changes visible in the
+  CMakeLists.txt file that called `add_subdirectory`. You can now add the new
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: You can now add the source file `Test_ProductionCodeFileBeingTested.cpp`.

File: docs/DevGuide/WritingTests.md [hunk 7]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+  source file, which should be named `Test_ProductionCodeFileBeingTest.cpp`.
+- You are adding a new directory:<br>
+  If the directory is a new lowest level directory you must add a
+  `add_subdirectory` call to `tests/Unit/CMakeLists.txt`. If it is a new
+  subdirectory you must add a `add_subdirectory` call to the
+  `CMakeLists.txt` file in the directory where you are adding the
+  subdirectory. Next you should read the port on adding a new test library.
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: port -> part

File: docs/DevGuide/WritingTests.md [hunk 8]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+- You are adding a new test library:<br>
+  After creating the subdirectory for the new test library you must add a
+  `CMakeLists.txt` file. See one `tests/Unit/DataStructures/CMakeLists.txt` for
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: See one -> See

File: docs/DevGuide/WritingTests.md [hunk 9]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+  an example of one. The `LIBRARY` and `LIBRARY_SOURCES` variables set the name
+  of the test library and the source files to be compiled into it. The library
+  name should be of the format `Test_ProductionLibraryName`, for example
+  `Test_DataStructures`. The library sources should be only the source files in
+  the current directory, additional source files in subdirectories can be added
+  to the same library. This is what the call to `add_subdirectory` in
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: Reword sentence This is what the call...

File: docs/DevGuide/WritingTests.md [hunk 10]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+function, which means that it should be followed by `{ /* test code */ }`. For
+example,
+\snippet Test_Tensor.cpp example_spectre_test_case
+From within a `SPECTRE_TEST_CASE` you are able to do all the things you would
+normally do in a C++ function, including calling other functions, setting
+variables, using lambdas, etc.
+
+The `CHECK` macro in the above example is provided by
+[Catch2](https://github.com/catchorg/Catch2) and is used to check conditions. We
+also a `CHECK_ITERABLE_APPROX` macro which checks if two iterable containers of
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: also a -> use a verb

File: docs/DevGuide/WritingTests.md [hunk 11]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+`double`s or two `double`s are approximately equal. `CHECK_ITERABLE_APPROX` is
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: reverse about `or`

File: docs/DevGuide/WritingTests.md [hunk 12]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+especially useful for comparing `Tensor`s, `DataVector`s, and
+`Tensor<DataVector>`s since it will iterate over nested containers as well.
+
+\warning Catch's `CHECK` statement only prints numbers out to approximately 10
+digits at most, so you should generally prefer `CHECK_ITERABLE_APPROX` for
+checking double precision numbers, unless you want to check that two numbers are
+bitwise identical.
+
+All unit tests must finish within a few seconds, the hard limit is 5, but having
+unit tests that long is strongly discouraged. The should typically complete in
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: `The should` -> They should

File: docs/DevGuide/WritingTests.md [hunk 13]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+less than half a second. Tests that are longer are often no longer testing a
+small enough unit of code and should either be split into several unit tests or
+moved to an integration test.
+
+#### Discovering New and Renamed Tests
+
+When you add a new test to a source file or rename an existing test the change
+needs to be discovered by the testing infrastructure. This is done by build the
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: build -> building

File: docs/DevGuide/WritingTests.md [hunk 14]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+target `rebuild_cache`, e.g. by running `make rebuild_cache`.
+
+#### Testing Pointwise Functions
+
+Pointwise functions should generally be tested in two different ways. The first
+is by taking input from an analytic solution and checking that the compute
+result is corrected. The second is to use the random number generation
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: corrected -> correct

File: docs/DevGuide/WritingTests.md [hunk 15]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+#### Discovering New and Renamed Tests
+
+When you add a new test to a source file or rename an existing test the change
+needs to be discovered by the testing infrastructure. This is done by build the
+target `rebuild_cache`, e.g. by running `make rebuild_cache`.
+
+#### Testing Pointwise Functions
+
+Pointwise functions should generally be tested in two different ways. The first
+is by taking input from an analytic solution and checking that the compute
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: compute -> computed

File: docs/DevGuide/WritingTests.md [hunk 16]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+result is corrected. The second is to use the random number generation
+comparison with python infrastructure. In this approach the C++ function being
+tested is re-implemented in python and the results are compared. If the function
+does sums over tensor indices then the einsum package should be used in python
+to provide an alternative implementation of the loop structure. The python
+implementations should be in a file name `TestFunctions.py` in the same
+directory as the `Test_*.cpp` source file and have the same name as the C++
+function being tested. It is possible to test C++ functions that return by value
+and ones that return by `gsl::not_null`. In the latter case, since it is
+possible to return multiple values one python function taking all
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: values -> values,

File: docs/DevGuide/WritingTests.md [hunk 17]
```diff
@@ -4,12 +4,166 @@ See LICENSE.txt for details.
...
+non-`gsl::not_null` arguments must be supplied for each `gsl::not_null` argument
+to the C++. To perform the test the `pypp::check_with_random_values()` must be
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md]: must -> function must

File: docs/DevGuide/WritingTests.md [hunk 18]
```diff
@@ -30,6 +184,10 @@ For example,
 #endif
 }
 ```
-
 If the `ifdef SPECTRE_DEBUG` is omitted then compilers will correctly flag
 the code as being unreachable which results in warnings.
+
+You can also test `ERROR`s inside your code. These tests need to have the
```
Comment (nilsdeppe) [inline docs/DevGuide/WritingTests.md:191]: add example of error test

**Review by kidder — DISMISSED** [review]

File: docs/MainSite/Main.md [hunk 19]
```diff
@@ -15,9 +15,97 @@ Table of Contents
...
+href="https://github.com/sxs-collaboration/spectre/fork"><img
+src="https://img.shields.io/badge/Fork%20%20on%20%20GitHub--blue.svg?style=social"
+alt="fork"
+data-canonical-src="https://img.shields.io/badge/Fork%20%20on%20%20GitHub--blue.svg?style=social"
+style="max-width:100%;"></a>
+
+<a
+href="https://github.com/sxs-collaboration/spectre"><img
+src="https://img.shields.io/badge/Star%20%20on%20%20GitHub--blue.svg?style=social"
+alt="fork"
```
Comment (wthrowe) [inline docs/MainSite/Main.md]: What is this thing for?  It is labeled "Star on GitHub", has alt="fork", and just links to the main SpECTRE GitHub page.

File: docs/DevGuide/WritingTests.md [hunk 20]
```diff
@@ -4,12 +4,167 @@ See LICENSE.txt for details.
...
+directory as the `Test_*.cpp` source file and have the same name as the C++
+function being tested. It is possible to test C++ functions that return by value
+and ones that return by `gsl::not_null`. In the latter case, since it is
+possible to return multiple values, one python function taking all
+non-`gsl::not_null` arguments must be supplied for each `gsl::not_null` argument
+to the C++. To perform the test the `pypp::check_with_random_values()` function
+must be called. For example, the following checks various C++ functions by
+calling into `pypp`:
+
+snippet Test_Pypp.cpp HALP
```
Comment (wthrowe) [inline docs/DevGuide/WritingTests.md]: Should there be an actual snippet here?

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: docs/MainSite/Main.md (hunk 19 above)
Comment (nilsdeppe) [reply docs/MainSite/Main.md]: Sorry, that should say "star", and from what I found there isn't a way to directly "Star". Maybe we should name this "View on GitHub" instead?

File: docs/DevGuide/WritingTests.md (hunk 20 above)
Comment (nilsdeppe) [reply docs/DevGuide/WritingTests.md]: Yes, was waiting for the python PR to get merged :)

**Comment by nilsdeppe** [comment]
Rebased on top of develop and updated the documentation at the link in the summary

**Comment by nilsdeppe** [comment]
~Ugh...~ Rebased on develop again

**Review by kidder — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Comment by nilsvu** [comment]
This should have been labeled _breaking change_ because of the renaming.

**Pull request merged**